lib/Dialect/SV/Transforms/SVExtractTestCode.cpp

```diff
@@ -334,21 +334,49 @@ void SVExtractTestCodeImplPass::runOnOperation() {
       if (auto mod = symCache.getDefinition(inst.moduleNameAttr()))
         if (mod->getAttr("firrtl.extract.assert.extra"))
           return true;
-    return isa<AssertOp>(op) || isa<FinishOp>(op) || isa<FWriteOp>(op) ||
-           isa<AssertConcurrentOp>(op);
+
+    // Error with a message which starts with "assert:" or "Assertion failed" is
+    // considered as an assert.
+    if (auto error = dyn_cast<ErrorOp>(op)) {
+      if (auto message = error.message())
+        return message.getValue().startswith("assert:") ||
+               message.getValue().startswith("Assertion failed");
+      return false;
+    }
+
+    // FWrite with a message which starts with "assert:" or "Assertion failed"
+    // is considered as assert.
+    if (auto fwrite = dyn_cast<FWriteOp>(op))
+      return fwrite.string().startswith("assert:") ||
+             fwrite.string().startswith("Assertion failed");
+
+    return isa<AssertOp>(op) || isa<FinishOp>(op) ||
+           isa<AssertConcurrentOp>(op) || isa<FatalOp>(op);
   };
+
   auto isAssume = [&symCache](Operation *op) -> bool {
     if (auto inst = dyn_cast<hw::InstanceOp>(op))
       if (auto mod = symCache.getDefinition(inst.moduleNameAttr()))
         if (mod->getAttr("firrtl.extract.assume.extra"))
           return true;
+
+    // FWrite with a message which starts with "assume:" is considered as
+    // assume.
+    if (auto fwrite = dyn_cast<FWriteOp>(op))
+      return fwrite.string().startswith("assume:");
```

Comment:
This may not be reachable due to these printf-encoded assumes being converted by the parser. Ditto for the cover extraction below.

Reply:
What's going on is that there are certain printf-encoded assumes/covers that are not being picked up during parsing and converted from printfs to `assume`/`cover`. This code is then picking up the pieces. I think it's reasonable to merge this with just the change to `ErrorOp` and then fix this in the parser for `assume:` and `cover:`.

Reply:
Out of curiosity, why isn't the parser picking them up?  Is it reasonable to fix the root issue there?

Reply:
Yeah I think the parser should replace this with an assert :thinking:. Could the parser actually already be picking it up, but then `LowerToHW` lowers this to `sv.if`, `sv.fwrite`, `sv.error`, `sv.stop`, which `SVExtractTestCode` doesn't know how to pick up?

Reply:
It is a bug of parser that printf ops with cover/assume messages are not converted into sv.cover/sv.assume. I think sv.error is only used for assertion of `ifElseFatal` format so I'll delete the modification to cover/assume here.

Reply:
Ah this is a great catch!

Reply:
Found a bug https://github.com/llvm/circt/issues/2384